icu4c/source/i18n/unitsdata.cpp

```diff
@@ -112,6 +320,44 @@ const ConversionRateInfo *ConversionRates::extractConversionInfo(StringPiece sou
     return nullptr;
 }
 
+U_I18N_API UnitPreferences::UnitPreferences(UErrorCode &status) {
+    LocalUResourceBundlePointer unitsBundle(ures_openDirect(NULL, "units", &status));
+    UnitPreferencesSink sink(&unitPrefs_, &metadata_);
+    ures_getAllItemsWithFallback(unitsBundle.getAlias(), "unitPreferenceData", sink, status);
+}
+
+void U_I18N_API
+UnitPreferences::getPreferencesFor(const char *category, const char *usage, const char *region,
+                                   MaybeStackVector<UnitPreference> *outPreferences,
+                                   UErrorCode &status) {
+    // UnitPreferenceMetadata *m = getMetadata(category, usage, region);
+    int32_t idx = getPreferenceIndex(&metadata_, category, usage, region, status);
+    if (U_FAILURE(status)) { return; }
+    if (idx < 0) { // Unnecessary if "status" was set correctly. TODO(review): should I delete this?
+                   // (Excessively "defensive"?)
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return;
+    }
+    UnitPreferenceMetadata *m = metadata_[idx];
+    for (int32_t pref = m->prefsOffset; pref < m->prefsOffset + m->prefsCount; pref++) {
+        UnitPreference *p = unitPrefs_[pref];
+        // TODO(review): we're making a full copy of the preferences here.
```

Comment:
Return a read-only slice of the `unitPrefs_` array: a pointer into `unitPrefs_` and a length.

```cpp
outLength = m->prefsCount;  // if outLength is a mutable reference
return unitPrefs_ + m->prefsOffset;
```

Reply:
This requires unitPrefs_ to be a contiguous and stable block of memory. MaybeStackVector doesn't document supporting it. MemoryPool is documented as being simple/minimalist. I'll read MaybeStackVector and MaybeStackArray code to find out if this would be reasonable/easy. If so - I'd need to document the "raw access" to the underlying "T stackArray[stackCapacity];"

I'm not sure we want to expose these classes' internals that much though - so perhaps I drop the MaybeStackArray and just do raw pointer management in UnitPreferences.

Reply:
As discussed, use the `const` keyword to guarantee that the MaybeStackVector won't be changing.

Reply:
MaybeStackVector<UnitPreference> has a MaybeStackArray<UnitPreference*> underlying it, and I think I can't get at the array-of-pointers via any of the accessors. (operator[] does a dereferencing.)

Reply:
My bad: there *is* a `T** getAlias()` I can use. Pardon the noise please. :)